doc/designs/quic-design/quic-ackm.md

```diff
@@ -0,0 +1,411 @@
+     */
+    unsigned int pkt_space :2;
+
+    /* 1 if the packet has one or more ACK-eliciting frames. */
+   unsigned int is_ack_eliciting :1;
+} QUIC_ACKM_RX_PKT;
+
+int QUIC_ACKM_on_rx_packet(QUIC_ACKM *ackm, const QUIC_ACKM_RX_PKT *pkt);
+```
+
+### On RX ACK Frame
+
+This must be called whenever an ACK frame is received. It should be called
+after any call to `QUIC_ACKM_on_rx_packet`.
+
+The ranges of packet numbers being acknowledged are passed as an argument.
+`pkt_space` is one of the `QUIC_PN_SPACE_*` values, specifying the packet number
+space of the containing packet. `rx_time` is the time the frame was
+received.
+
+This function causes `on_acked` callbacks to be invoked on applicable packets.
+
+Returns 1 on success.
+
+```c
+typedef struct quic_ackm_ack_range_st {
+    /*
+     * Represents an inclusive range of packet numbers [start, end]. 
+     * start must be <= end.
+     */
+    QUIC_PN start, end;
+} QUIC_ACKM_ACK_RANGE;
+
+typedef struct quic_ackm_ack {
+    /*
+     * A sequence of packet number ranges [[start, end]...].
+     *
+     * The ranges must be sorted in descending order, for example:
+     *      [ 95, 100]
+     *      [ 90,  92]
+     *      etc.
+     * 
+     * As such, ack_ranges[0].end is always the highest packet number
+     * being acknowledged and ack_ranges[num_ack_ranges-1].start is
+     * aalways the lowest packet number being acknowledged.
+     *
+     * num_ack_ranges must be greater than zero, as an ACK frame must
+     * acknowledge at least one packet number.
+     */
+    const QUIC_ACKM_ACK_RANGE  *ack_ranges;
+    size_t                      num_ack_ranges;
+
+    OSSL_TIME                   delay_time;
+    uint64_t                    ect0, ect1, ecnce;
+} QUIC_ACKM_ACK;
+
+int QUIC_ACKM_on_rx_ack_frame(QUIC_ACKM *ackm, const QUIC_ACKM_ACK *ack,
+                              int pkt_space, OSSL_TIME rx_time);
+```
+
+### On Packet Space Discarded
+
+This must be called whenever a packet number space is discarded. ACK-tracking
+information for the number space is thrown away. Any previously provided
+`QUIC_ACKM_TX_PKT` structures have their `on_discarded` callback invoked,
+providing an opportunity for them to be freed.
+
+Returns 1 on success.
+
+```c
+int QUIC_ACKM_on_pkt_space_discarded(QUIC_ACKM *ackm, int pkt_space);
+```
+
+### On Handshake Connfirmed
+
+This should be called by the user when the QUIC handshake is confirmed. The PTO
+algorithm behaves differently depending on whether the QUIC handshake is
+confirmed yet.
+
+Returns 1 on success.
+
+```c
+int QUIC_ACKM_on_handshake_confirmed(QUIC_ACKM *ackm);
+```
+
+### On Timeout
+
+This must be called whenever the loss detection deadline expires.
+
+```c
+int QUIC_ACKM_on_timeout(QUIC_ACKM *ackm);
+```
+
+Queries
+-------
+
+These functions allow information about the status of the ACK manager to be
+obtained.
+
+### Get Loss Detection Timeout
+
+This returns a deadline after which `QUIC_ACKM_on_timeout` should be called.
```

Comment:
I don't like this approach.  It means that the ACK manager is maintaining an internal list of packets in flight and keeping it sorted.  This is exactly what the timers were meant to be for.  Cut the middleman and use timers directly.

Reply:
Wait, what? The ACK manager necessarily has to keep a list of packets in flight. That's intrinsic to its operation: https://www.rfc-editor.org/rfc/rfc9002.html#name-on-sending-a-packet

The in-flight packets will be tracked as a linked list internally. I don't know of any implementation which doesn't keep this data.

Reply:
It was the _keeping it sorted_ that concerned me :)

Reply:
There will be a list of packets per PN space. This list will be sorted by PN and kept as a linked list. This ordering is needed by the ACK manager for timer-unrelated reasons. Since PNs  should monotonically increase in practice I'm not expecting to need to do a sorted insert but only ever append to the end of the list.

Reply:
Yeah, you'll probably keep the highest PN sent somewhere and assert if that changes.